### PR TITLE
Common BDN code refactor

### PR DESCRIPTION
**Pull request by adamsitnik**
I was not sure how to name the project, `BechmarkDotNet.Extensions` was my best idea.

### CODE REVIEW

File: src/benchmarks/micro/MicroBenchmarks.csproj [hunk 1]
```diff
@@ -32,6 +30,10 @@
     <PackageReference Include="ZeroFormatter.Analyzer" Version="1.1.1" />
   </ItemGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\..\harness\BenchmarkDotNet.Extensions\BenchmarkDotNet.Extensions.csproj" />
```
Comment (jorive) [inline src/benchmarks/micro/MicroBenchmarks.csproj]: Maybe the PATH could be something like `..\..\BenchmarkDotNet\Extensions\Extensions.csproj`?

File: src/benchmarks/micro/Program.cs [hunk 2]
```diff
@@ -18,24 +14,9 @@ class Program
         static int Main(string[] args)
             => BenchmarkSwitcher
                 .FromAssembly(typeof(Program).Assembly)
-                .Run(args, GetConfig())
+                .Run(args, RecommendedConfig.Create(
+                    artifactsPath: new DirectoryInfo(Path.Combine(Path.GetDirectoryName(typeof(Program).Assembly.Location), "BenchmarkDotNet.Artifacts")),
```
Comment (jorive) [inline src/benchmarks/micro/Program.cs]: This should probably be moved to the `RecommendedConfig` instead.

File: src/benchmarks/real-world/Microsoft.ML.Benchmarks/Microsoft.ML.Benchmarks.csproj [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
     <OutputType>Exe</OutputType>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="BenchmarkDotNet" Version="0.11.3" />
     <PackageReference Include="Microsoft.ML" Version="0.10.0" />
   </ItemGroup>
 
+  <ItemGroup>
+      <ProjectReference Include="..\..\..\harness\BenchmarkDotNet.Extensions\BenchmarkDotNet.Extensions.csproj" />
```
Comment (jorive) [inline src/benchmarks/real-world/Microsoft.ML.Benchmarks/Microsoft.ML.Benchmarks.csproj]: w.r.t. The referenced project PATH please see above.

File: src/benchmarks/real-world/Microsoft.ML.Benchmarks/Program.cs [hunk 4]
```diff
@@ -25,26 +20,11 @@ class Program
         static int Main(string[] args)
             => BenchmarkSwitcher
                 .FromAssembly(typeof(Program).Assembly)
-                .Run(args, GetConfig())
+                .Run(args, RecommendedConfig.Create(
+                    artifactsPath: new DirectoryInfo(Path.Combine(Path.GetDirectoryName(typeof(Program).Assembly.Location), "BenchmarkDotNet.Artifacts")),
```
Comment (jorive) [inline src/benchmarks/real-world/Microsoft.ML.Benchmarks/Program.cs]: Same as above.

File: src/harness/BenchmarkDotNet.Extensions/BenchmarkDotNet.Extensions.csproj [hunk 5]
```diff
@@ -0,0 +1,18 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (jorive) [inline src/harness/BenchmarkDotNet.Extensions/BenchmarkDotNet.Extensions.csproj]: Should we add:
```xml
<UseSharedCompilation>false</UseSharedCompilation>
<GenerateDocumentationFile>False</GenerateDocumentationFile>
<TreatWarningsAsErrors>True</TreatWarningsAsErrors>
<WarningLevel>4</WarningLevel>
<SuppressNETCoreSdkPreviewMessage>True</SuppressNETCoreSdkPreviewMessage>
<SignAssembly>false</SignAssembly>
```

**Comment by jorive** [comment]
These package versions could be updated:
```diff
-    <PackageReference Include="Jil" Version="2.15.4" />
+    <PackageReference Include="Jil" Version="2.16.0" />
-    <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
-    <PackageReference Include="protobuf-net" Version="2.3.7" />
-    <PackageReference Include="System.Memory" Version="4.5.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.1" />
+    <PackageReference Include="protobuf-net" Version="2.4.0" />
+    <PackageReference Include="System.Memory" Version="4.5.2" />
```

File: src/benchmarks/micro/Program.cs (hunk 2 above)
Comment (adamsitnik) [reply src/benchmarks/micro/Program.cs]: the problem is that we want `typeof(Program).Assembly.Location` which is different than `BenchmarkDotNet.Extensions.dll` location

We could use `Assembly.GetEntryAssembly` or ``Assembly.GetExecutingAssembly` but I have bad experience using those methods (they have platform specific behaviour depending on what has actually run the benchmarks) and AFAIR they are not implemented on CoreRT (which I hope to add sooon)

File: src/harness/BenchmarkDotNet.Extensions/BenchmarkDotNet.Extensions.csproj (hunk 5 above)
Comment (adamsitnik) [reply src/harness/BenchmarkDotNet.Extensions/BenchmarkDotNet.Extensions.csproj]: I added the reference to `common.props` file

File: src/benchmarks/real-world/Microsoft.ML.Benchmarks/Program.cs (hunk 4 above)
Comment (adamsitnik) [reply src/benchmarks/real-world/Microsoft.ML.Benchmarks/Program.cs]: same as above (`typeof(Program).Assembly` != `BenchmarkDotNet.Extensions.dll`)

File: src/benchmarks/micro/MicroBenchmarks.csproj (hunk 1 above)
Comment (adamsitnik) [reply src/benchmarks/micro/MicroBenchmarks.csproj]: I am not sure if this is a better idea. I don't think that we are ever gonna need more `BenchmarkDotNet` projects but I think that we might add some other harnesses in the future.

**Comment by adamsitnik** [comment]
> These package versions could be updated:

Yes, but it could affect the reported time. For example update of Newtonsoft.Json from 11 to 12 can have some perf impact.  We are now testing the stability of the reported results and I would prefer to not update them now.

**Comment by adamsitnik** [comment]
@jorive PTAL at the changes that I have introduced in https://github.com/dotnet/performance/pull/341/commits/ee7d5c7eb204d6571defbb058cee696cecf9fcf5

the CI has failed with some strange error after I have introduced a new project:

```log
$ pushd "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\artifacts\bin\BenchmarkDotNet.Extensions\Release\netcoreapp3.0"
Searching BenchmarkDotNet output files with: BenchmarkDotNet.Artifacts/**/*-full.json
$ popd
$ pushd "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\artifacts\bin"
$ popd
Traceback (most recent call last):
  File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchmarks_ci.py", line 250, in <module>
    __main(sys.argv[1:])
  File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchmarks_ci.py", line 245, in __main
    benchview.run_scripts(args, verbose, BENCHMARKS_CSPROJ)
  File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchview.py", line 454, in run_scripts
    jobtype=args.benchview_run_type
  File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchview.py", line 167, in submission
    raise ValueError("No `measurement.json` were specified.")
ValueError: No `measurement.json` were specified.
```

we tried to search for the results in `bin\BenchmarkDotNet.Extensions\Release\netcoreapp3.0` instead `
bin\MicroBenchmarks\Release\netcoreapp3.0`

I debugged it and it turned out that we were searching for the results in the **first** folder that matched `'**/{Configuration}/{TargetFramework}` 

I have changed it to `**/{ProjectName}/{Configuration}/{TargetFramework}`.

**Comment by jorive** [comment]
> @jorive PTAL at the changes that I have introduced in ee7d5c7
> the CI has failed with some strange error after I have introduced a new project:
> $ pushd "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\artifacts\bin\BenchmarkDotNet.Extensions\Release\netcoreapp3.0"
> Searching BenchmarkDotNet output files with: BenchmarkDotNet.Artifacts/**/*-full.json
> $ popd
> $ pushd "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\artifacts\bin"
> $ popd
> Traceback (most recent call last):
>   File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchmarks_ci.py", line 250, in <module>
>     __main(sys.argv[1:])
>   File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchmarks_ci.py", line 245, in __main
>     benchview.run_scripts(args, verbose, BENCHMARKS_CSPROJ)
>   File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchview.py", line 454, in run_scripts
>     jobtype=args.benchview_run_type
>   File "C:\dotnetbuild\work\2ad47326-3428-4085-b5f4-99a575a381e5\Payload\scripts\benchview.py", line 167, in submission
>     raise ValueError("No `measurement.json` were specified.")
> ValueError: No `measurement.json` were specified.
> 
> we tried to search for the results in bin\BenchmarkDotNet.Extensions\Release\netcoreapp3.0 instead  bin\MicroBenchmarks\Release\netcoreapp3.0
> I debugged it and it turned out that we were searching for the results in the first folder that matched '**/{Configuration}/{TargetFramework}
> I have changed it to **/{ProjectName}/{Configuration}/{TargetFramework}.

Sure, I'll take a look.

**Comment by adamsitnik** [comment]
This has blocked me in #352. I am going to merge this now to unblock #352 and allow for ML.NET per runs.

**Comment by jorive** [comment]
@adamsitnik Did you solve/fix the exception you were getting?

**Comment by adamsitnik** [comment]
@jorive yes, I was able to repro it locally, fix and test

**Pull request merged**